From 3ccc3437abcc484a9c34d65b1b2fa5f4d252acce Mon Sep 17 00:00:00 2001
From: Simmi Pateriya <simmip@codeaurora.org>
Date: Mon, 26 Nov 2012 23:06:01 +0530
Subject: [PATCH] ASoC: wcd9310: Avoid button polling noise on second insertion

If the headset is inserted before HPH path tears
down,there is noise heard due to button polling.
Micbias is switched to VDDIO to avoid this noise.

Signed-off-by: Simmi Pateriya <simmip@codeaurora.org>
(cherry picked from commit 213750140c754a3b22dedf28d61951e68785e2db)

Change-Id: Ia781e6a60a37c7ca9b1c9cb08d8bb9a8fda3f4bf
Signed-off-by: Sridhar Gujje <sgujje@codeaurora.org>
---
 sound/soc/codecs/wcd9310.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/sound/soc/codecs/wcd9310.c b/sound/soc/codecs/wcd9310.c
index a6694df..c074e2e 100644
--- a/sound/soc/codecs/wcd9310.c
+++ b/sound/soc/codecs/wcd9310.c
@@ -1,4 +1,4 @@
-/* Copyright (c) 2011-2012, Code Aurora Forum. All rights reserved.
+/* Copyright (c) 2011-2013, The Linux Foundation. All rights reserved.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License version 2 and
@@ -6410,6 +6410,13 @@ void tabla_find_plug_and_report(struct snd_soc_codec *codec,
 		 * only report the mic line
 		 */
 		tabla_codec_report_plug(codec, 1, SND_JACK_HEADSET);
+		if (!tabla->mbhc_micbias_switched &&
+			tabla_is_hph_pa_on(codec)) {
+			/*If the headphone path is on, switch the micbias
+			to VDDIO to avoid noise due to button polling */
+			tabla_codec_switch_micbias(codec, 1);
+			pr_debug("%s: HPH path is still up\n", __func__);
+		}
 		msleep(100);
 		tabla_codec_start_hs_polling(codec);
 	} else if (plug_type == PLUG_TYPE_HIGH_HPH) {
@@ -6785,7 +6792,6 @@ static void tabla_codec_detect_plug_type(struct snd_soc_codec *codec)
 	} else if (plug_type == PLUG_TYPE_HEADSET) {
 		pr_debug("%s: Headset detected\n", __func__);
 		tabla_codec_report_plug(codec, 1, SND_JACK_HEADSET);
-
 		/* avoid false button press detect */
 		msleep(50);
 		tabla_codec_start_hs_polling(codec);
@@ -7195,6 +7201,7 @@ static void tabla_hs_gpio_handler(struct snd_soc_codec *codec)
 					    0x08, 0x00);
 			/* Turn off override */
 			tabla_turn_onoff_override(codec, false);
+			tabla_codec_switch_micbias(codec, 0);
 		}
 	}
 
